router/h2/src/e2e/scala/io/buoyant/router/h2/FlowControlEndToEndTest.scala

```diff
@@ -0,0 +1,131 @@
+import io.buoyant.test.FunSuite
+import scala.collection.mutable.ListBuffer
+
+class FlowControlEndToEndTest
+    extends FunSuite
+    with ClientServerHelpers {
+
+  override val logLevel = Level.OFF
+
+  test("client/server request flow control") {
+    val streamP = new Promise[Stream]
+    val server = Downstream.mk("server") { req =>
+      streamP.setValue(req.data)
+      Response(Status.Ok, Stream.Nil)
+    }
+    val client = upstream(server.server)
+    try {
+      val writer = Stream()
+      val req = Request("http", Method.Get, "host", "/path", writer)
+      val rsp = await(client(req))
+      assert(rsp.status == Status.Ok)
+      testFlowControl(reader(await(streamP)), writer)
+    } finally {
+      await(client.close())
+      await(server.server.close())
+    }
+  }
+
+  test("client/server response flow control") {
+    val writer = Stream()
+    val server = Downstream.mk("server") { _ => Response(Status.Ok, writer) }
+    val client = upstream(server.server)
+    try {
+      val req = Request("http", Method.Get, "host", "/path", Stream.Nil)
+      val rsp = await(client(req))
+      assert(rsp.status == Status.Ok)
+      testFlowControl(reader(rsp.data), writer)
+    } finally {
+      await(client.close())
+      await(server.server.close())
+    }
+  }
+
+  val WindowSize = 65535
+
+  def testFlowControl(reader: Stream.Reader, writer: Stream.Writer[Frame]) = {
+    // Put two windows' worth of data on the stream
+    val release0, release1 = new Promise[Unit]
+    def releaser(p: Promise[Unit]) = () => {
+      p.setDone()
+      Future.Unit
+    }
+
+    // The first frame is too large to fit in a window. It should not
+    // be released until all of the data has been flushed.  This
+    // cannot happen until the reader reads and erleases some of the
+    // data.
+    val frame0 = Frame.Data(mkBuf(WindowSize + 1024), false, releaser(release0))
+    val frame1 = Frame.Data(mkBuf(WindowSize - 1024), true, releaser(release1))
+    assert(!release0.isDefined && !release1.isDefined)
+    log.debug("offering 2 frames with %d", 2 * WindowSize)
+    assert(writer.write(frame0))
+    assert(writer.write(frame1))
+    assert(!release0.isDefined && !release1.isDefined)
+
+    // Read a full window, without releasing anything.
+    var read = 0
+    val frames = ListBuffer.empty[Frame.Data]
+    while (read < WindowSize) {
+      await(reader.read()) match {
+        case _: Frame.Trailers => fail("unexpected trailers")
+        case d: Frame.Data =>
+          frames += d
+          read += d.buf.length
+      }
+    }
+    assert(frames.map(_.buf.length).sum == read)
+    assert(read == WindowSize)
+    assert(!release0.isDefined && !release1.isDefined)
+
+    // At this point, we've read an entire window of data from
+    // the stream without releasing any of it.  Subsequent reads
+    // should not complete until data is released.
+    val rf0 = reader.read()
+    for (_ <- 0 to 10000) assert(!rf0.isDefined)
```

Comment:
It's not reliable-reliable, but how do you test a negative...  This is really the inverse of `eventually`.  This is my hacky version of `notEventually`